Add assistant prefill for chat templates and TextGenerationPipeline

Rocketknight1 opened the pull request:
Something that's been requested several times both internally and on Github is assistant prefill: The ability to begin the model's response for it and let it continue.

We use a slightly hacky solution suggested by @Narsil, but which I think will work in all cases that I know of: When we do assistant prefill, we simply truncate the formatted chat to the end of the final message text, removing any tokens that indicate the end of an assistant message, which will cause the model to simply continue from wherever the final message ends.

This PR also updates `TextGenerationPipeline` so that if you pass a chat where the final message is an assistant message, it will assume that you want to treat the final message as an assistant prefill. Before this PR, it would start a new message if you did this, which would generally result in malformed output because most models were not trained with multiple consecutive assistant messages.

Fixes #33096

Fixes #32213

Rocketknight1 committed:
Add assistant prefill to chat templates

Rocketknight1 committed:
Add assistant prefill to pipeline

Rocketknight1 committed:
Add assistant prefill to pipeline

LysandreJik commented:
Let me know when you'd like a review @Rocketknight1!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_33198). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

Rocketknight1 committed:
Tweak another test that ended in assistant message

Rocketknight1 committed:
Update tests that ended in assistant messages

Rocketknight1 committed:
Update tests that ended in assistant messages

Rocketknight1 commented:
@LysandreJik should be ready for review now! I had to tweak a couple of other tests that ended in assistant messages - the output was garbage in those cases anyway, and the tests were just checking that it wasn't changing unexpectedly.

For visibility:
cc @philschmid who requested this. Cc @Narsil for TGI and @xenova for `transformers.js`, though note that this doesn't change Jinja behaviour at all, it just adds a flag that trims the output string in Python! Also cc @gante for generation.

Fixes #33096

Fixes #32213

DIYer22 commented:
Instead of `assistant_prefill`, may just add a new parameter called `prefill_last_message` to `apply_chat_template`. I feel like there's no need to limit prefill to the assistant role. This way, it's compatible with the current needs and can also adapt to potential future needs. Plus, the parameter name and the corresponding changes are simple and clear enough that there won't be any misunderstandings.

Rocketknight1 commented:
@DIYer22 I'm not sure about that - the "assistant" role is universal across all chat models that I know of. Although some users might want the model to continue a user or system message, I think that's more of a fun/research thing than a common use-case, and I think it's okay to expect people to construct their own prompts/templates for those cases.

Rocketknight1 committed:
Replace assistant_prefill with continue_final_message

Rocketknight1 committed:
Allow passing continue_final_message to pipeline

Rocketknight1 commented:
Hmn, okay, there's more demand for it than I thought - you were right @DIYer22!

@Tostino @DIYer22 I've replaced `assistant_prefill` with `continue_final_message`. It has the same behaviour as `assistant_prefill`, but will no longer raise an error if the final message is not an `assistant` message.

I've kept the `TextGenerationPipeline` behaviour the same, however - since that's aimed at simpler use-cases! It will assume that users want a prefill only if the final message is an `assistant` message.

Rocketknight1 committed:
Small fixup

Rocketknight1 committed:
Add continue_final_message as a pipeline kwarg

Rocketknight1 committed:
Update docstrings

Rocketknight1 commented:
Update: Pipeline now correctly handles the `continue_final_message` kwarg

Rocketknight1 commented:
@LysandreJik sorry for the delay, should be ready for actual review now!

LysandreJik commented on src/transformers/pipelines/text_generation.py:
```diff
@@ -391,8 +413,19 @@ def postprocess(self, model_outputs, return_type=ReturnType.FULL_TEXT, clean_up_
                         # Explicit list parsing is necessary for parsing chat datasets
-                        all_text = list(prompt_text.messages) + [{"role": "assistant", "content": all_text}]
-
+                        if continue_final_message is None:
+                            continue_final_message = prompt_text.messages[-1]["role"] == "assistant"
```
Maybe add a comment explaining what is being done under the hood here (so assuming that the request is to continue the final message if the last message was already sent by the assistant?)

Rocketknight1 replied:
Done!

LysandreJik commented on src/transformers/tokenization_utils_base.py:
```diff
@@ -1737,10 +1738,14 @@ def apply_chat_template(
                 argument, as the model's template will be used by default.
-            add_generation_prompt (bool, *optional*): Whether to end the prompt with the token(s) that indicate
-                the start of an assistant message. This is useful when you want to generate a response from the model.
+            add_generation_prompt (bool, *optional*): If this is set, a prompt with the token(s) that indicate
+                the start of an assistant message will be appended to the formatted output. This is useful when you want to generate a response from the model.
                 Note that this argument will be passed to the chat template, and so it must be supported in the
                 template for this argument to have any effect.
+            continue_final_message (bool, *optional*): If this is set, the chat will be formatted so that the final
+                message in the chat is open-ended, without any EOS tokens. The model will continue this message
+                rather than starting a new one. This allows you to "prefill" part of
+                the model's response for it. Cannot be used at the same time as `add_generation_prompt`.
```
For consistency with the rest of the docs (line returns)
```suggestion
            add_generation_prompt (bool, *optional*): 
                If this is set, a prompt with the token(s) that indicate
                the start of an assistant message will be appended to the formatted output. This is useful when you want to generate a response from the model.
                Note that this argument will be passed to the chat template, and so it must be supported in the
                template for this argument to have any effect.
            continue_final_message (bool, *optional*): 
                If this is set, the chat will be formatted so that the final
                message in the chat is open-ended, without any EOS tokens. The model will continue this message
                rather than starting a new one. This allows you to "prefill" part of
                the model's response for it. Cannot be used at the same time as `add_generation_prompt`.
```

LysandreJik commented on tests/pipelines/test_pipelines_text_generation.py:
```diff
@@ -191,6 +189,68 @@ def test_small_chat_model_pt(self):
+        # Here we check that passing a chat that ends in an assistant message is handled correctly
+        # by continuing the final message rather than starting a new one
+        text_generator = pipeline(
+            task="text-generation", model="rocketknight1/tiny-gpt2-with-chatml-template", framework="pt"
```
Can you move that model to `hf-internal-testing`? We're trying to move away from user/contributor/maintainer-hosted checkpoints in our CI

Rocketknight1 replied:
Done!

LysandreJik commented on tests/pipelines/test_pipelines_text_generation.py:
```diff
@@ -191,6 +189,68 @@ def test_small_chat_model_pt(self):
+        # Here we check that passing a chat that ends in an assistant message is handled correctly
+        # by continuing the final message rather than starting a new one
+        text_generator = pipeline(
+            task="text-generation", model="rocketknight1/tiny-gpt2-with-chatml-template", framework="pt"
```
same here

Rocketknight1 replied:
Done!

LysandreJik reviewed (approved):
Ok makes sense! Thanks @Rocketknight1

Rocketknight1 committed:
Move repos to hf-internal-testing!

Rocketknight1 committed:
Update src/transformers/tokenization_utils_base.py

Co-authored-by: Lysandre Debut <hi@lysand.re>

Rocketknight1 committed:
Add explanatory comment

Rocketknight1 committed:
make fixup

Rocketknight1 commented:
@LysandreJik all comments addressed/merged, and moved all the repos to `hf-internal-testing`. Also updated a couple of other cases where I was using personal repos for tests that were already merged to `main`, and moves those too.

Rocketknight1 committed:
Update chat templating docs to explain continue_last_message

LysandreJik reviewed (approved):
Looks good! Thanks for iterating on this @Rocketknight1

Tostino commented:
Thank you very much for the work on this @Rocketknight1. I'm really happy you came up with a solution that helps solve all the needs brought up.

Rocketknight1 commented:
Thanks @Tostino! Please let me know if you encounter any issues while using it - you can try it out right now by installing from `main`.

Tostino commented:
@Rocketknight1 There seems to be an issue...

I was testing it the other day and just couldn't get it working...and then I needed to drive my continue on my roadtrip. Just got back to testing and realized there is a discrepancy between the code and the documentation.

Documentation says [`continue_last_message`](https://huggingface.co/docs/transformers/main/en/chat_templating#what-does-continuelastmessage-do)  and code is actually `continue_final_message`.

Rocketknight1 commented:
@Tostino I have no idea how I overlooked that! I think I just changed the name of the argument quite late in the PR, opening a fix now.
